### PR TITLE
Jettison vxsort support in the GC when optimizing for size

### LINKED ISSUE #84749: Is vxsort with workstation GC a common scenario?
We currently enable VXSORT in GC under following ifdefs:

https://github.com/dotnet/runtime/blob/6875ba02dbde6477abe16f0863348e617cc03fcb/src/coreclr/gc/gc.cpp#L21-L25

I experimentally modified the `ifdef` by adding ` && defined(SERVER_GC)`. With that, the size of a selfcontained NativeAOT hello world went down by ~10%. We currently build two flavors for NativeAOT runtime - one with the server GC and one with the workstation GC. The server GC is opt in through the publicly documented `ServerGarbageCollection` MSBuild property and not the default.

|                        | VXSORT  | No VXSORT |
|------------------------|---------|-----------|
| Default                | 1.41 MB |   1.30 MB |
| InvariantGlobalization | 1.19 MB |   1.08 MB |

(It is likely the InvariantGlobalization case will be the default for projects created with `dotnet new console -aot`.)

This feels like a pretty good saving. The question is - would this be a compromise? Is having WKS GC with heaps large enough that we would use VXSORT a common scenario? We want to have comparable perf with Native AOT, but also size is one of the things we take seriously.

Cc @dotnet/gc @dotnet/ilc-contrib

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Vxsort comes with precomputed tables that are more than 100 kB in size.

Jettison them when optimizing for size. Selfcontained hello world with optimize for size + invariant globalization is now 1.08 MB (a 10% saving).

Fixes #84749.

Cc @dotnet/ilc-contrib @dotnet/gc

### CODE REVIEW

**Comment by MichalStrehovsky** [comment]
> "in the GC" in general or only when it's compiled for WKS GC?

In general, both server and wks GC. 

Note that will be done only when the user says they're willing to trade throughput for size (i.e. it will require a gesture from the user that spells out optimization preference is size). Disabled VXSORT will not be the default.

> how does one get these disabled by default things enabled in NativeAoT? is there a page that explains that?

There will be a doc page on the optimization preference setting, but it will probably not document the exact mechanism. It will only say we'll do whatever's reasonable to get size down at the cost of throughput. E.g. we align method bodies differently with this setting as well.

**Comment by cshung** [comment]
Instead of just turning it off, can we build the table at runtime (maybe after startup) instead?

**Comment by jkotas** [comment]
> Instead of just turning it off, can we build the table at runtime (maybe after startup, outside of GC pause) instead?

IIRC, the vxsort static footprint is dominated by code. The supporting data tables are relatively small.

**Review by jkotas — APPROVED** [review]

**Comment by VSadov** [comment]
If I need a server GC, since I expect lots of concurrency and some short-lived allocations, but overall do not expect an enormous heap, then I can pass `IlcDisableVxsort=true`?  
 
It looks so from the change, but a bit hard to follow, - the 100 kB of vxsort stuff will not even be linked into the app, right?

**Comment by MichalStrehovsky** [comment]
> If I need a server GC, since I expect lots of concurrency and some short-lived allocations, but overall do not expect an enormous heap, then I can pass IlcDisableVxsort=true?

This change doesn't depend on whether we're building with server/workstation GC. That approach was rejected based on discussion in #84749.

We key off of two things: are we optimizing for size? (the (planned to be) publicly documented switch), or someone specified `IlcDisableVxsort` (undocumented switch).

Then there's a bunch of extra conditions because vxsort only works on x64 Windows and bifurcating the build further based on whether control flow guard is enabled on not didn't seem worth it so we don't disable it if control flow guard is enabled.
